File: graph-algorithms/graph_algorithms.f90
! Graph Algorithms in Modern Fortran
!
! Implementations of fundamental graph algorithms:
! - Breadth-First Search (BFS)
! - Depth-First Search (DFS)
! - Dijkstra's Shortest Path
!
! Graph representation: Adjacency Matrix (simpler for Fortran)

program graph_algorithms
    implicit none

    integer, parameter :: MAX_V = 100
    integer, parameter :: INF = 999999

    ! Test graph
    integer, parameter :: n_vertices = 6
    integer :: graph(n_vertices, n_vertices)
    integer :: i, j

    print '(A)', repeat('=', 70)
    print '(A)', '                GRAPH ALGORITHMS IN FORTRAN'
    print '(A)', repeat('=', 70)
    print *

    ! Initialize graph with INF (no edges)
    graph = INF
    do i = 1, n_vertices
        graph(i, i) = 0  ! Distance to self is 0
    end do

    ! Add edges (undirected graph with weights)
    ! Edge (0,1) weight 4
    graph(1, 2) = 4
    graph(2, 1) = 4

    ! Edge (0,2) weight 3
    graph(1, 3) = 3
    graph(3, 1) = 3

    ! Edge (1,2) weight 1
    graph(2, 3) = 1
    graph(3, 2) = 1

    ! Edge (1,3) weight 2
    graph(2, 4) = 2
    graph(4, 2) = 2

    ! Edge (2,3) weight 4
    graph(3, 4) = 4
    graph(4, 3) = 4

    ! Edge (3,4) weight 2
    graph(4, 5) = 2
    graph(5, 4) = 2

    ! Edge (4,5) weight 6
    graph(5, 6) = 6
    graph(6, 5) = 6

    ! Print graph
    call print_graph(graph, n_vertices)
    print *

    ! Test BFS
    print '(A)', 'Test 1: Breadth-First Search (BFS)'
    print '(A)', repeat('-', 70)
    call bfs(graph, n_vertices, 1)  ! Start from vertex 1 (0-indexed as 1 in Fortran)
    print *

    ! Test DFS
    print '(A)', 'Test 2: Depth-First Search (DFS)'
    print '(A)', repeat('-', 70)
    call dfs(graph, n_vertices, 1)
    print *

    ! Test Dijkstra
    print '(A)', 'Test 3: Dijkstras Shortest Path Algorithm'
    print '(A)', repeat('-', 70)
    call dijkstra(graph, n_vertices, 1)

    print *
    print '(A)', repeat('=', 70)
    print '(A)', 'Key Points:'
    print '(A)', '- BFS: O(V + E) time, explores level by level'
    print '(A)', '- DFS: O(V + E) time, explores depth-first'
    print '(A)', '- Dijkstra: O(V²) time with array, O((V+E)logV) with heap'
    print '(A)', '- Adjacency matrix: O(V²) space, O(1) edge lookup'
    print '(A)', repeat('=', 70)

contains

    ! ===================================================================
    ! GRAPH UTILITIES
    ! ===================================================================

    subroutine print_graph(graph, n)
        integer, intent(in) :: graph(:,:), n
        integer :: i, j

        print '(A)', 'Graph adjacency matrix:'
        print '(A)', '(INF = no edge)'
        print *

        ! Print header
        write(*, '(A)', advance='no') '   '
        do j = 1, n
            write(*, '(I6)', advance='no') j-1
        end do
        print *

        ! Print matrix
        do i = 1, n
            write(*, '(I2, A)', advance='no') i-1, ': '
            do j = 1, n
                if (graph(i, j) == INF) then
                    write(*, '(A6)', advance='no') 'INF'
                else
                    write(*, '(I6)', advance='no') graph(i, j)
                end if
            end do
            print *
        end do
    end subroutine print_graph

    ! ===================================================================
    ! BREADTH-FIRST SEARCH (BFS)
    ! ===================================================================

    subroutine bfs(graph, n, start)
        integer, intent(in) :: graph(:,:), n, start
        logical :: visited(n)
        integer :: queue(n), front, rear
        integer :: current, i

        visited = .false.
        front = 1
        rear = 1
        queue(rear) = start

        visited(start) = .true.

        write(*, '(A, I0, A)', advance='no') 'BFS traversal from vertex ', start-1, ': '

        do while (front <= rear)
            current = queue(front)
            front = front + 1

            write(*, '(I0, A)', advance='no') current-1, ' '

            ! Visit all adjacent vertices
            do i = 1, n
                if (graph(current, i) /= INF .and. graph(current, i) /= 0 &
                    .and. .not. visited(i)) then
                    visited(i) = .true.
                    rear = rear + 1
                    queue(rear) = i
                end if
            end do
        end do

        print *
    end subroutine bfs

    ! ===================================================================
    ! DEPTH-FIRST SEARCH (DFS)
    ! ===================================================================

    subroutine dfs(graph, n, start)
        integer, intent(in) :: graph(:,:), n, start
        logical :: visited(n)

        visited = .false.

        write(*, '(A, I0, A)', advance='no') 'DFS traversal from vertex ', start-1, ': '
        call dfs_util(graph, n, start, visited)
        print *
    end subroutine dfs

    recursive subroutine dfs_util(graph, n, vertex, visited)
        integer, intent(in) :: graph(:,:), n, vertex
        logical, intent(inout) :: visited(:)
        integer :: i

        visited(vertex) = .true.
        write(*, '(I0, A)', advance='no') vertex-1, ' '

        ! Visit all adjacent vertices
        do i = 1, n
            if (graph(vertex, i) /= INF .and. graph(vertex, i) /= 0 &
                .and. .not. visited(i)) then
                call dfs_util(graph, n, i, visited)
            end if
        end do
    end subroutine dfs_util

    ! ===================================================================
    ! DIJKSTRA'S SHORTEST PATH
    ! ===================================================================

    subroutine dijkstra(graph, n, src)
        integer, intent(in) :: graph(:,:), n, src
        integer :: dist(n)
        logical :: visited(n)
        integer :: i, count, min_dist, u, v

        ! Initialize
        dist = INF
        dist(src) = 0
        visited = .false.

        do count = 1, n
            ! Find minimum distance vertex not yet visited
            min_dist = INF
            u = -1

            do i = 1, n
                if (.not. visited(i) .and. dist(i) < min_dist) then
                    min_dist = dist(i)
                    u = i
                end if
            end do

            if (u == -1) exit  ! All remaining vertices are unreachable

            visited(u) = .true.

            ! Update distances of adjacent vertices
            do v = 1, n
                if (.not. visited(v) .and. graph(u, v) /= INF &
                    .and. dist(u) /= INF &
                    .and. dist(u) + graph(u, v) < dist(v)) then
                    dist(v) = dist(u) + graph(u, v)
                end if
            end do
        end do

        ! Print results
        print *
        print '(A, I0)', 'Dijkstras shortest paths from vertex ', src-1
        print '(A)', 'Vertex    Distance from Source'
        print '(A)', repeat('-', 40)

        do i = 1, n
            write(*, '(I4, A)', advance='no') i-1, '         '
            if (dist(i) == INF) then
                print '(A)', 'INF'
            else
                print '(I0)', dist(i)
            end if
        end do
    end subroutine dijkstra

end program graph_algorithms

File: graph-algorithms/graph_algorithms.R
# Graph Algorithms in R
#
# Implementations of fundamental graph algorithms:
# - Breadth-First Search (BFS)
# - Depth-First Search (DFS)
# - Dijkstra's Shortest Path
#
# Graph representation: Adjacency List (efficient for sparse graphs)

#' Create an empty graph
#'
#' @param num_vertices Number of vertices
#' @return Graph as a list of adjacency lists
create_graph <- function(num_vertices) {
    graph <- vector("list", num_vertices)
    for (i in 1:num_vertices) {
        graph[[i]] <- list()
    }
    return(graph)
}

#' Add edge to graph
#'
#' @param graph Graph structure
#' @param from Source vertex (1-indexed)
#' @param to Destination vertex (1-indexed)
#' @param weight Edge weight
#' @param directed Whether graph is directed
add_edge <- function(graph, from, to, weight = 1, directed = FALSE) {
    # Add edge from -> to
    graph[[from]] <- c(graph[[from]], list(list(vertex = to, weight = weight)))

    # For undirected graph, add edge to -> from
    if (!directed) {
        graph[[to]] <- c(graph[[to]], list(list(vertex = from, weight = weight)))
    }

    return(graph)
}

#' Print graph structure
print_graph <- function(graph) {
    cat("Graph adjacency list:\n")
    for (v in 1:length(graph)) {
        cat(sprintf("Vertex %d:", v - 1))
        for (neighbor in graph[[v]]) {
            cat(sprintf(" -> %d(w:%d)", neighbor$vertex - 1, neighbor$weight))
        }
        cat("\n")
    }
}

# ===================================================================
# BREADTH-FIRST SEARCH (BFS)
# ===================================================================

#' Breadth-First Search
#'
#' @param graph Graph structure
#' @param start Starting vertex (1-indexed)
#' @return Vector of visited vertices in BFS order
bfs <- function(graph, start) {
    num_vertices <- length(graph)
    visited <- rep(FALSE, num_vertices)
    queue <- c()
    result <- c()

    # Enqueue start vertex
    visited[start] <- TRUE
    queue <- c(queue, start)

    while (length(queue) > 0) {
        # Dequeue
        current <- queue[1]
        queue <- queue[-1]

        result <- c(result, current - 1)  # Store 0-indexed

        # Visit all adjacent vertices
        for (neighbor in graph[[current]]) {
            adj_vertex <- neighbor$vertex
            if (!visited[adj_vertex]) {
                visited[adj_vertex] <- TRUE
                queue <- c(queue, adj_vertex)
            }
        }
    }

    return(result)
}

# ===================================================================
# DEPTH-FIRST SEARCH (DFS)
# ===================================================================

#' Depth-First Search (recursive)
#'
#' @param graph Graph structure
#' @param start Starting vertex (1-indexed)
#' @return Vector of visited vertices in DFS order
dfs <- function(graph, start) {
    num_vertices <- length(graph)
    visited <- rep(FALSE, num_vertices)
    result <- c()

    dfs_util <- function(vertex) {
        visited[vertex] <<- TRUE
        result <<- c(result, vertex - 1)  # Store 0-indexed

        # Visit all adjacent vertices
        for (neighbor in graph[[vertex]]) {
            adj_vertex <- neighbor$vertex
            if (!visited[adj_vertex]) {
                dfs_util(adj_vertex)
            }
        }
    }

    dfs_util(start)
    return(result)
}

# ===================================================================
# DIJKSTRA'S SHORTEST PATH
# ===================================================================

#' Dijkstra's Shortest Path Algorithm
#'
#' @param graph Graph structure
#' @param src Source vertex (1-indexed)
#' @return Vector of shortest distances from source
dijkstra <- function(graph, src) {
    num_vertices <- length(graph)
    dist <- rep(Inf, num_vertices)
    visited <- rep(FALSE, num_vertices)

    dist[src] <- 0

    for (count in 1:num_vertices) {
        # Find minimum distance vertex not yet visited
        min_dist <- Inf
        u <- -1

        for (v in 1:num_vertices) {
            if (!visited[v] && dist[v] < min_dist) {
                min_dist <- dist[v]
                u <- v
            }
        }

        if (u == -1 || dist[u] == Inf) break  # All reachable vertices processed

        visited[u] <- TRUE

        # Update distances of adjacent vertices
        for (neighbor in graph[[u]]) {
            v <- neighbor$vertex
            weight <- neighbor$weight

            if (!visited[v] && dist[u] + weight < dist[v]) {
                dist[v] <- dist[u] + weight
            }
        }
    }

    return(dist)
}

# ===================================================================
# MAIN DEMONSTRATION
# ===================================================================

cat(rep('=', 70), '\n')
cat('                    GRAPH ALGORITHMS IN R\n')
cat(rep('=', 70), '\n\n')

# Create a graph with 6 vertices
num_vertices <- 6
graph <- create_graph(num_vertices)

# Add edges (undirected graph with weights)
# Using 1-indexed vertices in R
graph <- add_edge(graph, 1, 2, weight = 4)
graph <- add_edge(graph, 1, 3, weight = 3)
graph <- add_edge(graph, 2, 3, weight = 1)
graph <- add_edge(graph, 2, 4, weight = 2)
graph <- add_edge(graph, 3, 4, weight = 4)
graph <- add_edge(graph, 4, 5, weight = 2)
graph <- add_edge(graph, 5, 6, weight = 6)

# Print graph structure
print_graph(graph)
cat('\n')

# Test BFS
cat('Test 1: Breadth-First Search (BFS)\n')
cat(rep('-', 70), '\n')
bfs_result <- bfs(graph, 1)  # Start from vertex 0 (index 1 in R)
cat(sprintf('BFS traversal from vertex 0: %s\n', paste(bfs_result, collapse = ' ')))
cat('\n')

# Test DFS
cat('Test 2: Depth-First Search (DFS)\n')
cat(rep('-', 70), '\n')
dfs_result <- dfs(graph, 1)  # Start from vertex 0 (index 1 in R)
cat(sprintf('DFS traversal from vertex 0: %s\n', paste(dfs_result, collapse = ' ')))
cat('\n')

# Test Dijkstra
cat('Test 3: Dijkstra\'s Shortest Path Algorithm\n')
cat(rep('-', 70), '\n')
distances <- dijkstra(graph, 1)  # Start from vertex 0 (index 1 in R)

cat('\nDijkstra\'s shortest paths from vertex 0:\n')
cat('Vertex\tDistance from Source\n')
cat(rep('-', 40), '\n')

for (i in 1:num_vertices) {
    if (is.infinite(distances[i])) {
        cat(sprintf('%d\t\tINF\n', i - 1))
    } else {
        cat(sprintf('%d\t\t%d\n', i - 1, distances[i]))
    }
}

cat('\n', rep('=', 70), '\n')
cat('Key Points:\n')
cat('- BFS: O(V + E) time, explores level by level\n')
cat('  * Shortest path in unweighted graphs\n')
cat('  * Uses queue data structure\n')
cat('- DFS: O(V + E) time, explores depth-first\n')
cat('  * Good for detecting cycles, connectivity\n')
cat('  * Uses recursion/stack\n')
cat('- Dijkstra: O((V + E) log V) with heap, O(V²) with array\n')
cat('  * Shortest path in weighted graphs (non-negative weights)\n')
cat('  * Uses priority queue/min-heap\n')
cat('- Adjacency list: O(V + E) space, efficient for sparse graphs\n')
cat(rep('=', 70), '\n')

# Visualize paths (simple text representation)
cat('\nPath visualization:\n')
cat('Graph structure (vertex -> neighbors):\n')
cat('0 -> 1(4), 2(3)\n')
cat('1 -> 0(4), 2(1), 3(2)\n')
cat('2 -> 0(3), 1(1), 3(4)\n')
cat('3 -> 1(2), 2(4), 4(2)\n')
cat('4 -> 3(2), 5(6)\n')
cat('5 -> 4(6)\n')
cat('\nShortest paths from vertex 0:\n')
cat('0 -> 0: 0 (self)\n')
cat('0 -> 1: 4 (direct)\n')
cat('0 -> 2: 3 (direct)\n')
cat('0 -> 3: 6 (via 1 or 2)\n')
cat('0 -> 4: 8 (via 3)\n')
cat('0 -> 5: 14 (via 4)\n')
